token-lending/program/src/dex_market.rs

```diff
@@ -0,0 +1,304 @@
+//! Dex market used for simulating trades
+
+use crate::{error::LendingError, math::Decimal, state::Reserve};
+use arrayref::{array_refs, mut_array_refs};
+use serum_dex::critbit::Slab;
+use solana_program::{account_info::AccountInfo, program_error::ProgramError, pubkey::Pubkey};
+use std::{cell::RefMut, collections::VecDeque, convert::TryFrom};
+
+/// Side of the dex market order book
+#[derive(PartialEq)]
+enum Side {
+    Bid,
+    Ask,
+}
+
+/// Input currency for trade simulator
+#[derive(Clone, Copy, PartialEq)]
+enum Currency {
+    Base,
+    Quote,
+}
+
+/// The role of the trader
```

Comment:
Borrow amount can be specified in either deposited collateral or desired loan amount but the simulation used to compute the unknown should be equivalent in either direction. So if we simulate filling asks (let's say buying SOL with USDC) as the taker, then the other direction must simulate by fillings asks as the maker (by selling SOL for USDC).

Reply:
Definitely agree that they need to go in both directions.  This was a terminology nit, since in the calculations we're always acting as a `Taker` and never as a `Maker`.  We're always looking to fill the trade immediately, and not putting an order into the book to be taken by someone else.

Reply:
Fixed!